File: 0883. Projection Area of 3D Shapes/0883.java
// time complexity: O(n^2)
// space complexity: O(1)
public class Solution {
    // iterative loops
    public int projectionArea(int[][] grid) {
        int n = grid.length;
        int area = 0;
        // top & front projections in one pass over rows
        for (int i=0; i<n; i++) {
            int maxRow = 0;
            for (int j=0; j<n; j++) {
                if (grid[i][j] > 0) {
                    // top view: each non-zero stack casts a 1×1 shadow
                    area++;
                }
                // track maximum height in this row for front view
                maxRow = Math.max(maxRow, grid[i][j]);
            }
            // add row-max for front projection
            area += maxRow;
        }
        // side projection: max of each column
        for (int j=0; j<n; j++) {
            int maxCol = 0;
            for (int i=0; i<n; i++) {
                maxCol = Math.max(maxCol, grid[i][j]);
            }
            area += maxCol;
        }
        return area;
    }
}

File: 0883. Projection Area of 3D Shapes/0883.py
# time complexity: O(n^2)
# space complexity: O(1)
class Solution(object):
    def projectionArea(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        n = len(grid)
        top = 0
        front = 0
        side = 0

        # compute top (xy) and front (yz) projections
        for i in range(n):
            max_row = 0
            for j in range(n):
                v = grid[i][j]
                # top view: each non-zero tower contributes 1
                if v > 0:
                    top += 1
                # track the tallest tower in this row for front view
                if v > max_row:
                    max_row = v
            front += max_row

        # compute side (zx) projection by scanning columns manually
        for j in range(n):
            max_col = 0
            for i in range(n):
                v = grid[i][j]
                # track the tallest tower in this column
                if v > max_col:
                    max_col = v
            side += max_col

        return top + front + side